### DISCUSSION TITLE
Is it a good idea to keep import map in Github and use it inside Kubernetes as Config Map

### DISCUSSION THREAD

**Discussion by ajasingh** [single-spa/import-map-deployer#115] [Q&A]
Right now we are keeping the Import-map in a EFS which is mounted inside the Node.js service , but right now there is no track who update import-map  and also difficult to  rollback to a previous version of import-map easily , so we were thinking is it a good idea to keep import-map in a git and update the file in git whenever a new version is deployed so that we have clear history of changes and the production release can be easily achieved using tags

**Comment by internettrans** [comment]
I discussed this a couple years ago with a company - generally I think that rollbacks and versioning are best handled by CI/CD tools rather than in import-map-deployer.

However, CI / CD tools generally are per-repo, and the overall import map is modified by the CI processes of many repo. So rolling back the entire import map is not possible via CI / CD. For that reason, I think it could be good to add history / rollback functionality to import-map-deployer.

I think the main options for doing so are below:

1. git
2. database
3. object versioning in S3 / object store.


### git

Git is nice for an infrastructure-as-code (IAC) approach, and has built-in audit trail. However, generally with IAC the code repository is the source of truth and changes to that repository result in deployments. In that case, the import-map-deployer server is not necessary, as the CI/CD process for the repo containing the import maps for each environment can perform the update without risk of race condition. I view that approach as a completely viable, but different, approach from import-map-deployer. Not better or worse. However, I don't think that using git as a mirror of production, or only as an audit trail, makes as much sense, since if pushes to that repo originated outside of import-map-deployer, the git repo would be out of sync with production.

### db

A database holding the audit trail is a nice approach, except that requiring users running import-map-deployer to run an extra DB could be an adoption hurdle. Organizations have preferences about pg/mysql/mongo/dynamo/etc, and also have preferences about whether dbs are run in docker containers or right in the cloud infrastructure (usually built-in cloud db service like RDS is preferred). Additionally, a problem of running the db in the docker container is that docker container volumes are generally ephemeral and not synced between instances of import-map-deployer. Updating the import-map-deployer image would likely cause an erasure of the audit trail for many users who don't have everything properly configured.

### object versioning

Object versioning in the cloud provider is an appealing option, as it fits the paradigm of import-map-deployer, where the live file in S3/Azure/GCP/DO is the source of truth. Here's the documentation for s3: https://docs.aws.amazon.com/AmazonS3/latest/userguide/Versioning.html. I'm most interested in this option being added to import-map-deployer, although I haven't used object versioning before so I'd need to make sure it really would meet the needs of keeping an audit trail and allowing rollbacks. The downside of the approach is that it's unclear who would call the import-map-deployer API for rolling back - would it be a CI process for a repo? A human debugging production issues? I think those are solvable things, and am interested in this approach being investigated further.

What are your thoughts about this?

**Reply by ajasingh** [reply]
Using cloud tools is a limitation to us and introducing db might be a hassle so personally now we are looking at simple Node.js service which will keep the import-map in the git and update it on the fly

**Reply by internettrans** [reply]
> we are looking at simple Node.js service which will keep the import-map in the git and update it on the fly

This is problematic since the git repo can become out of sync with production, since other users besides import-map-deployer could push. Making it a code repo invites developers to push to it, which would often result in confusion about why pushing to the repo doesn't fix production.

**Reply by ajasingh** [reply]
We will restrict the writes to only automation user so that no one manually can update it from any place any time do you still see any problem happening like this ?

**Reply by internettrans** [reply]
You will change the settings, but other users of import-map-deployer won't. Like I mentioned in my original post, putting the import map in Git hints at an infrastructure as code approach, but this wouldn't be infrastructure as code since the code doesn't determine the state of production.

For these reasons, I do not think that this is a good solution to add generally to import-map-deployer. Feel free to do so for your organization, but I think one of the other approaches described above would be better for general use with import-map-deployer.
